packages/wrangler/src/d1/execute.tsx

```diff
@@ -348,41 +346,194 @@ async function executeRemotely({
 		accountId,
 		name
 	);
-	if (preview && !db.previewDatabaseUuid) {
-		const error = new UserError(
-			"Please define a `preview_database_id` in your wrangler.toml to execute your queries against a preview database"
-		);
-		logger.error(error.message);
-		throw error;
+	if (preview) {
+		if (!db.previewDatabaseUuid) {
+			const error = new UserError(
+				"Please define a `preview_database_id` in your wrangler.toml to execute your queries against a preview database"
+			);
+			logger.error(error.message);
+			throw error;
+		}
+		db.uuid = db.previewDatabaseUuid;
 	}
-	const dbUuid = preview ? db.previewDatabaseUuid : db.uuid;
-	logger.log(`🌀 Executing on remote database ${name} (${dbUuid}):`);
+	logger.log(
+		`🌀 Executing on ${
+			db.previewDatabaseUuid ? "preview" : "remote"
+		} database ${name} (${db.uuid}):`
+	);
 	logger.log(
 		"🌀 To execute on your local development database, remove the --remote flag from your wrangler command."
 	);
 
-	const results: QueryResult[] = [];
-	for (const sql of batches) {
-		if (multiple_batches)
-			logger.log(
-				chalk.gray(`  ${sql.slice(0, 70)}${sql.length > 70 ? "..." : ""}`)
-			);
+	if (input.file) {
+		// TODO: do we need to update hashing code if we upload in parts?
+		const etag = await md5File(input.file);
+
+		logger.log(
+			chalk.gray(
+				`Note: if the execution fails to complete, your DB will return to its original state and you can safely retry.`
+			)
+		);
+
+		const initResponse = await d1ApiPost<
+			ImportInitResponse | ImportPollingResponse | PollingFailure
+		>(accountId, db, "import", { action: "init", etag });
+
+		// An init response usually returns a {filename, uploadUrl} pair, except if we've detected that file
+		// already exists and is valid, to save people reuploading. In which case `initResponse` has already
+		// kicked the import process off.
+		const uploadRequired = "uploadUrl" in initResponse;
+		if (!uploadRequired) logger.log(`🌀 File already uploaded. Processing.`);
+		const firstPollResponse = uploadRequired
+			? await uploadAndBeginIngestion(
+					accountId,
+					db,
+					input.file,
+					etag,
+					initResponse
+			  )
+			: initResponse;
+
+		const finalResponse = await pollUntilComplete(
+			firstPollResponse,
+			accountId,
+			db
+		);
```

Comment:
Can we split this two functions out a bit more? Right now the upload function also requests the first polling response—it would be nice to have a function that _just_ did upload, and then handoff to a function that _just_ did polling?

Reply:
Sadly the API conflates the two a bit in the interests of performance. The "upload" endpoint does actually return a "polling response", which can also be "complete" (pretty common for things like migrations). The server side actually adds a small delay (1s currently) to increase the chance of things happening in a single round-trip.

You only need to "pollUntilComplete" if the first response isn't "complete", so I'll add a quick comment to clarify. But this was the best way I could figure out to capture the _actual_ behaviour of the API, rather than the logical order of operations...